### PR TITLE
TextField/PasswordField - Right button overlaps with text and some style and UX changes to Right Button

### PR DESCRIPTION
**Pull request by Exopandora**
Essentially the same changes as https://github.com/JFXtras/jfxtras-styles/pull/195

> Edit: Proposed changes
> 
> 1. Add right padding to Textfield so that text doesn't get under right button
> 2. Have pressed button color be the accent color with the graphic turning white
> 3. Only clear text when mouse released (instead of mouse pressed) -> Unify text clearing interaction with password display interaction

  
  

Focused and unfocused:
focused1
focused2

Pressed:
pressed

Personally i would additionally change the password textfield right button to be visible all the time to hide the padding (see the bottom textfield in figure 1). I would also disable the text selection while holding the password button.

### CODE REVIEW

**Comment by Exopandora** [comment]
Alternatively the background of the right button could be removed again

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css [hunk 1]
```diff
@@ -47,23 +47,31 @@ TextField {
...
+.text-field > .right-button > .right-button-graphic {
     -fx-shape: "M221.738,305.873l6.135,6.16l-2.875,2.863l-6.135-6.159l-6.263,6.237l-2.864-2.875l6.263-6.238l-6.177-6.202l2.875-2.863l6.177,6.201l6.244-6.22l2.864,2.876L221.738,305.873z";
 
     -fx-padding: 0.416667em;
-    -fx-background-color: #333;
+    -fx-background-color: #111;
 }
 
-TextField > .right-button:hover {
+.text-field > .right-button {
```
Comment (dukke) [inline FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:61]: We can't change this to ".text-field". There are reasons why "Textfield" is being used, namely there is a corner case where this will be a problem (check comment on base.css of JMetro).

Now that I'm looking at this, the comment on JMetro that says why we should use "TextField" should probably be added here as well to prevent regressions.

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css [hunk 2]
```diff
@@ -47,23 +47,31 @@ TextField {
     -fx-skin: "com.pixelduke.control.skin.FXTextFieldSkin";
 }
 
-TextField {
-    -right-button-visible: true;
```
Comment (dukke) [inline FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:51]: We need to have the -right-button-visible css property here. This is also for documentation reasons, so that users know this css property exists and that the default value is true.

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css [hunk 3]
```diff
@@ -47,23 +47,31 @@ TextField {
...
+.text-field {
+    -fx-padding: 0.3333333em 3em 0.3333333em 0.5833333em;
```
Comment (dukke) [inline FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:51]: why do we need to change the padding of the textfield?

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css [hunk 4]
```diff
@@ -47,23 +47,31 @@ TextField {
...
     -fx-padding: 0.416667em;
-    -fx-background-color: #333;
+    -fx-background-color: #111;
 }
 
-TextField > .right-button:hover {
+.text-field > .right-button {
     -fx-cursor: default;
+
+    -fx-background-color: transparent;
```
Comment (dukke) [inline FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:64]: Do we need to set this background color as transparent? What's the different to the original code which didn't have this?

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css [hunk 5]
```diff
@@ -47,23 +47,31 @@ TextField {
...
+    -fx-background-insets: -0.3333333em -0.6666666em -0.3333333em -0.5833333em; /* 4 8 4 7 px in em */
+    -fx-background-radius: 0 3 3 0, 2, 2;
 }
 
-TextField > .right-button:hover > .right-button-graphic {
-    -fx-background-color: #666666;
```
Comment (dukke) [inline FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:66]: Why are we removing this?

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css [hunk 6]
```diff
@@ -79,35 +87,10 @@ TextField > .right-button:hover > .right-button-graphic {
 .password-field {
     -right-button-visible: true;
 }
-.password-field > .right-button {
-    -fx-padding: 0.166667em 0.083333em 0.166667em 0.166667em;
-}
```
Comment (dukke) [inline FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:84]: Why does this need to get removed?

**Review by dukke — COMMENTED** [review]
There are some really nice additions that you are proposing here. I'm going to edit the description of this issue to detail all proposed changes and so that it is crystal clear what this PR is about.

I see though that there are also other changes in this PR that fall outside the scope of what we've discussed. I think we should try to stick to only changing what has been discusses so that it is easier to evaluate the PR.

There's also a risk of changing stuff without discussing it first. JMetro (and FXSkins which was derived from JMetro) has evolved over several years, some of the code present here has reasons to exist that may not make sense. However some of it is there to fix bugs that were discovered previously and changing it back may introduce unwanted regressions (example: changing `TextField` to `.text-field`).
Bottom line we should try to only change what has been discussed. If we want to change more stuff, it's better to discuss it first because some of it may be there for some reason (fix bugs) and also so because its easier for discussions sake to stick to only making changes that have been proposed previously.

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css (hunk 1 above)
Comment (Exopandora) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:61]: I changed it to `.text-field` beause it also applies to the password textfield. If it does cause issues we need to copy code.

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css [hunk 7]
```diff
@@ -47,23 +47,31 @@ TextField {
...
+    -fx-padding: 0.3333333em 3em 0.3333333em 0.5833333em;
 }
 
-TextField > .right-button > .right-button-graphic {
+.text-field > .right-button > .right-button-graphic {
     -fx-shape: "M221.738,305.873l6.135,6.16l-2.875,2.863l-6.135-6.159l-6.263,6.237l-2.864-2.875l6.263-6.238l-6.177-6.202l2.875-2.863l6.177,6.201l6.244-6.22l2.864,2.876L221.738,305.873z";
 
     -fx-padding: 0.416667em;
-    -fx-background-color: #333;
+    -fx-background-color: #111;
```
[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:58]: So it matches the color of the password textfield.

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css [hunk 8]
```diff
@@ -47,23 +47,31 @@ TextField {
...
 }
 
-TextField > .right-button:hover {
+.text-field > .right-button {
     -fx-cursor: default;
+
+    -fx-background-color: transparent;
+    -fx-background-insets: -0.3333333em -0.6666666em -0.3333333em -0.5833333em; /* 4 8 4 7 px in em */
```
[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:65]: Those are negative insets, so they actually expand the button. Otherwise it would look like this:
insets

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css (hunk 3 above)
Comment (Exopandora) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:51]: This prevents the text from going behind the right button. (The padding values are from the default jfx theme (Caspian) except the value for the right side)

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css (hunk 4 above)
Comment (Exopandora) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:64]: I dont know if its needed but the old documentation said that it needs a color in order to react to user input.

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css (hunk 2 above)
Comment (Exopandora) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:51]: This may have been an oversight, it can be added back

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css (hunk 5 above)
Comment (Exopandora) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:66]: This was removed to make the button more in line with the password textfield. The button only changes when you press (and hold) the button.

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css (hunk 6 above)
Comment (Exopandora) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:84]: Its now the same as the normal textfield right button.

File: FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java [hunk 9]
```diff
@@ -205,7 +205,7 @@ protected void layoutChildren(double x, double y, double w, double h) {
 
         rightButton.resize(clearButtonWidth, h);
         positionInArea(rightButton,
-                (x+w) - clearButtonWidth, y,
+                (x+w) + textField.snappedLeftInset() + textField.snappedRightInset() - (h + textField.snappedTopInset() + textField.snappedBottomInset()), y,
```
[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java:208]: We need the right inset because we set it to a higher value, so the text no longer overlaps. Otherwise it would look like this:
noinsets
Explanation:
explanation
We want to calculate the x position of z. The formula for this is `textfield_x + full_width_textfield - full height-textfield`. I would also not remove any other insets from the calculation because if another user sets custom insets the button may be positioned wrongly.

**Comment by dukke** [comment]
About the clear button / password button being visible all the time... I understand the problem and I agree it doesn't look good to have the text appear truncated.

I don't also like to have the right button always visible though. It makes the interface look more cluttered, I think?

The ideal solution would be to have the padding and have the text not be truncated, though I think that would only be possible with a considerable amount of effort.

Perhaps we could leave it as it is? Have the right button only appear when focused and have the text appear truncated when it reaches the limit (not ideal but perhaps the best solution at the moment)?

And potentially later we can come up with a solution for this (have the text not be truncated and only show the right button on hover)?

**Comment by Exopandora** [comment]
> The ideal solution would be to have the padding and have the text not be truncated, though I think that would only be possible with a considerable amount of effort.

Maybe we can cheat a bit. If we load the textfield we store the right inset value separately and set the value to the left value (and disable padding). When the text is non-empty we set the stored value back (and enable padding).

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css (hunk 2 above)
Comment (Exopandora) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:51]: Added back in 48cb59e

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css (hunk 4 above)
Comment (Exopandora) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css]: Removed in 201901c

**Comment by dukke** [comment]
> > The ideal solution would be to have the padding and have the text not be truncated, though I think that would only be possible with a considerable amount of effort.
> 
> Maybe we can cheat a bit. If we load the textfield we store the right inset value separately and set the value to the left value (and disable padding). When the text is non-empty we set the stored value back (and enable padding).

Didn't quite get what you mean...

Now that I'm thinking about it, perhaps we can simply have the textinput padding return to it's normal padding (modena padding) when it's not focused (i.e. doesn't have the `focused` pseudoclass). That would be easy to implement.

**Comment by Exopandora** [comment]
Thats actually quite a lot simpler

**Comment by dukke** [comment]
Sorry to bother you with this but now that we're at it, could you possibly add this comment to the beginning of the stylesheet definitions for TextField (right before the first `TextField` css rule):
```
/* Issue #149 (JMetro): Need to use TextField because ColorPicker has custom textfields with "text-field" styleclass that aren't actually of type
   TextField which will cause an IllegalArgumentException when creating the impl.jfxtras.styles.jmetro.TextFieldSkin object and passing it,
   for instance, the WebColorField */
```

Just so that we avoid future regressions because we forget this needs to be this way..

**Comment by dukke** [comment]
Sorry for being a bit of a pain reviewing your PR.. I hope you understand, just making sure we continue to maintain a high standard. Though some of my comments could be obviously wrong, just making sure we do indeed do things right by discussing them well.

**Comment by dukke** [comment]
Hi @Exopandora 
I was wondering if you need any help or don't agree with something, since it's been a while since you last posted comments?

Thanks again!

**Comment by Exopandora** [comment]
I thought there was nothing to do as i commented everything and changed anything that really needed a change.

**Review by dukke — COMMENTED** [review]
> I thought there was nothing to do as i commented on everything and changed anything that really needed a change.

I've left a few comments here in there on some of your changes..

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css [hunk 10]
```diff
@@ -88,6 +88,10 @@ TextField > .right-button:pressed > .right-button-graphic {
 .password-field {
     -right-button-visible: true;
 
+    -fx-padding: 0.3333333em 0.5833333em 0.3333333em 0.5833333em;
```
[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:95]: We already apply padding to the (normal) textfield (line 57). But we apply the one with the bigger right padding all the time as the right button is always visible.

File: FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java:208]: The way i thought:
The button should be a square (=> `width = height`). The easiest way to calculate x position of the button is to take the entire width of the textfield and subtract the width. In this case `width = height` so the x position of the button is `xpos_button = xpos_textfield + width_textfield - width_button = xpos_textfield + width_textfield - height_button`

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css (hunk 7 above)
[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:64]: In my option it looks better when its uniform as the buttons also serve kinda the same purpose. And when we are already at it we can also unify the experience.

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css (hunk 8 above)
[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:70]: The problem is that the stackpane does not grow

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css (hunk 5 above)
Comment (Exopandora) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:66]: I mean we could add hover back. But in my opinion they should either both have a hover effect or none. And my recurring example of the windows settings search bar also does not have a hover effect

**Comment by Exopandora** [comment]
Oh i didn't see those, sorry

File: FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (dukke) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java:208]: It could be rectangular. It depends on the dimensions the programmer defines for the whole textfield. In any way the button would need to cover the height of the textfield, so could end up being rectangular (i.e. non square)

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css (hunk 7 above)
[the remark this answers is not in the document]
Comment (dukke) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:64]: OK, we can keep that..

File: FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java:208]: I don't quite get that. If i set the width and the height of the button to the same value it must be a square.

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css (hunk 5 above)
Comment (dukke) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:66]: That's a good point. In that case I would prefer to also add the hover effect on the passwordfield. We don't need to make an exact copy of Microsoft fluent design, you'll see that JMetro doesn't copy it in several places (where I don't like the way fluent design looks or its UX). One example: Fluent Design Spinners don't look so good in my opinion and JMetro has a completely different look for them.
JMetro is inspired by FD but doesn't pretend to be a copy of it. 
Also FXSkins is a general purpose library for adding better skins to existing apps. JMetro will use it but so will other apps not using JMetro. My idea is that by default FXSkins looks good with Modena (the default javafx theme).

Comment (Exopandora) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:66]: I get that. But do you agree to add them back on both?

File: FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (dukke) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java:208]: I mean.. this is a general purpose library for adding better skins to existing apps. 
The programmer might decide to make textfields height taller by changing their padding values in that case the button should grow to fill the textfield height making it possible for the button to be rectangular (not square).
The idea here would be for the button to cover the textfield height, not for it to be restricted to being a square, I think.

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css (hunk 5 above)
Comment (dukke) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:66]: So, as you suggested, and I think that's a good idea: let's also add the hover effect to the password field "right button". Sound good?

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css (hunk 10 above)
[the remark this answers is not in the document]
Comment (dukke) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:95]: For consistency sake should we have the same right padding? The same right padding for textfield and passwordfield?

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css (hunk 8 above)
[the remark this answers is not in the document]
Comment (dukke) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:70]: I think that, probably, if we change the layoutchidlren method so that the button covers the textfield height we won't need to be changing the button insets to make it appear bigger.. 
As long as we position and resize the button to have the appropriate dimensions I think there won't be any need for any additional tweaks to make it bigger.

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css (hunk 10 above)
[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:95]:
```css
TextField {
    ...
    /*
    This is the same padding as the .password-field:focused
    We always need to apply it because the X-Button is always visible when there is text.
    */
    -fx-padding: 0.3333333em 3em 0.3333333em 0.5833333em; 
}

.password-field {
    ...
    /*
    We remove the right padding when the password-field is not focused because the button is no longer visible and the text can extend to the entire width of the password-field.
    */
    -fx-padding: 0.3333333em 0.5833333em 0.3333333em 0.5833333em;
}

.password-field:focused {
    /*
    This is the same padding as the TextField
    We always need to apply it because the button is visible when the password-field is focused.
    */
    -fx-padding: 0.3333333em 3em 0.3333333em 0.5833333em;
}
```

[the remark this answers is not in the document]
Comment (dukke) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:95]: Ops sorry, wrong comment (sleeped very little today 😖). 
What I meant to say is whether we should also remove the right padding when the Textfield is not focused as we do for the passwordfield?
This is what happens if I add only "x" to the upper textfield and then bring the focus to another control (the bottom textfield):
image

The last x appears cut. So text can get cut right now if we fill the textfield and then put the focus on another control.

File: FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css (hunk 8 above)
[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/resources/com/pixelduke/control/skin/fx-skins.css:78]: I checked again and you are right on this. It can be removed if we position it right

File: FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java:208]: I have a better solution in the working that should fix the issues. But i encountered the following issue:
`clearButtonWidth` does not actually calculate the width of the graphic. It always returns the same value.
I could bypass this by adding a minimum width to the right button. This would also lead to the nice sideeffect that the textfield right-button and the password-textfield right-button have the same width (if possible due to textfield dimensions).

[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java:208]: Current code is as follows:
```java
final double clearGraphicWidth = snapSizeX(rightButtonGraphic.prefWidth(-1));
final double clearButtonWidth = rightButton.snappedLeftInset() + clearGraphicWidth + rightButton.snappedRightInset();
        
rightButton.resize(clearButtonWidth, textField.getHeight());
positionInArea(rightButton, textField.getWidth() - clearButtonWidth, 0, clearButtonWidth, textField.getHeight(), 0, HPos.LEFT, VPos.TOP);
```

I still need to adjust the width according to the textfield dimensions

**Comment by Exopandora** [comment]
The width of the right button is now defined by the css for each right-button by `-fx-padding: 0 1.0833333em 0 1.0833333em`. This bypasses the issue that the different graphics do not return the same prefWidth. This also makes the width configurable by the end user. Additionally the button is now suitable for other aspect ratios. The default width is the default height of a textfield.

File: FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (dukke) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: Have you checked the comment I posted for the possible size, x and y definitions for the right button?
```
x = (x + w - textField.snappedRightInset()) - clearButtonWidth;
y = y + textField.snappedTopInset()
width = clearButtonWidth
height = h - (textField.snappedTopInset() + textField.snappedBottomInset())
```

The `prefWidth` method should give you the width of the right button according to what's defined in the css, etc..

Some comments to the code of your last comment that you posted:
1. I think you forgot to factor in the padding of the textfield (i.e. the insets). 
2. You don't need to get the height of the textfield: `textfield.getHeight()` and the width, they are already passed on to `layoutChildren` method.
3. You're resizing the button twice by calling `resize` and then `positionInArea` (`positionInArea` resizes and sets the x and y position).

[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: I tested your code and it does not place the button correct at all. `prefWidth` returns `0` for the password-textfield and `10.4` is returned for the normal textfield.

Regarding 1+2:
Printout for a simple textfield: `h`: 17.6 `textfield.getHeight()`: 25.6
=> `h != textfield.getHeight()`. I think that `textfield.getHeight()` takes the padding/insets into account and `h` does not

Regarding 3:
If we do not resize the button it will look like this:
grafik

[the remark this answers is not in the document]
Comment (dukke) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: 1 - There must be something else happening about `prefWidth` not working. I've been working with JavaFX since it's very first release (about 10/15 years ago) non stop, I've always relied on `prefWidth`, `prefHeight` for layout code. It should return a valid value unless there is something wrong with the size definitions we are setting for it in css, etc or something else.
One thing we could change when calling `rightButtonGraphic.prefWidth(-1)` is pass in the height, i.e: `rightButtonGraphic.prefWidth(h)` since we already know the height we want for the right button (we only pass -1 when we don't know the height).

2 - Yeah I think you're right., height and width passed might already be factoring in the padding of the textfield. 
Anyways, we don't need to be calling textfield.getHeight(), etc we can just use the width, height passed in to the layoutChildren method which are the dimension we should be using for laying out our nodes (they are the dimensions of the area that is provided for us to layout our nodes in the layoutChildren method).

3 - `resize`, sets the button width and height. `positionInArea` does a similar job: resizes and positions. 
If you think there is some issue with `positionInArea`, we could call `resize` (to set width and height) and then `setLayoutX` and `setLayoutY` (to position in the x and y axis). So that we don't repeat the same computation..

[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: Regarding 3: `resize` was used even before my PR

[the remark this answers is not in the document]
Comment (dukke) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: 3 - Yeap, you're right. My mistake, not yours, I don't think I should have been calling resize and positionInArea, I don't think that makes logical sense? As I was repeating the same computation twice.

(some of this code is quite old, JMetro has been around since 2013 if I'm not mistaken, some of it might be outdated or wrong as through the years I've learned more about JavaFX, etc)

Anyways sorry to bother you with some of my past mistakes. I think as we're re-implementing this code we might as well fix some of this mistakes.

[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: So you are saying i should somehow get rid of `resize`

[the remark this answers is not in the document]
Comment (dukke) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: You can call `resize` and then `setLayoutX` and `setLayoutY ` that way we don't repeat the "resize" part.
There's also the `resizeRelocate(...)` method, that we could use, which does both things in one go (setting width, height and x and y).

[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: The only thing i can imagine that could cause `prefWidth` to return 0 is `-fx-scale-shape: false;` for the password-textfield. The is that we cant remove it because otherwise the graphic is not being drawn (or scaled very wrongly).

[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: Currently im using the padding to adjust the width of the button (which is configurable by the user)

[the remark this answers is not in the document]
Comment (dukke) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: I just tried calling in your code in the `layoutChildren` method: 
``` 
double rightButtonWidth = rightButton.prefWidth(-1);
System.out.println("Right button width = " + rightButtonWidth);
```
And it prints `36`

There might be some detail that escaped you...?

However, since we know the height I think we should actually calling it this way:
```
double rightButtonWidth = rightButton.prefWidth(h);
```

[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: You are right with `rightButton.prefWidth(-1)` but i meant `rightButtonGraphic.prefWidth(-1)`.
```
TextField:
Right button width = 36.4
Right button graphic width = 10.0
password-textfield:
Right button width = 26.0
Right button graphic width = 0.0
```
The problem (at least i think) is that `rightButton.prefWidth` depends on `rightButtonGraphic.prefWidth` which returns DIFFERENT values for `TextField` (X-Icon) and `password-textfield` (Eye-Icon).

[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: `rightButton.prefWidth(h)` returns the following (when the padding is removed again):

```
TextField: 10.4
password-textfield: 0.0
```

They should be the same

[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: It works if i do `rightButton.setPrefWidth` in `layoutChildren`

[the remark this answers is not in the document]
Comment (dukke) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: The right button graphic is returning `0` for the passwordfield because we haven't defined a padding value for the passwordfield right button graphic in CSS.

[the remark this answers is not in the document]
Comment (dukke) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: > It works if i do `rightButton.setPrefWidth` in `layoutChildren`

It doesn't sound right to set a `prefWidth` in the `layoutChildren` method. 
The right button graphic is returning 0 for the passwordfield because we haven't defined a padding value for the passwordfield right button graphic in CSS.

If we want the width of the right button to be exactly the same for the passwordfield and the textfield then we can play with setting padding values on rightButton/rightButtonGraphic in CSS for the passwordfield and textfield.
It also has the added bonus of programmers being able to define whatever width they want for the right button by defining their own css rules.

[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: But the problem is that if i add padding the graphic scales differently

How its supposed to look:
grafik

With increased padding (from 0.416667 to 0.8em):
grafik

With increased padding (from 0.416667 to 0.8em) and `-fx-scale-shape: false` (just like the `password-textfield` icon):
grafik

[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]:
```
-fx-scale-x: 0.65;
-fx-scale-y: 0.65;
-fx-scale-z: 0.65;
```
would fix it actually

[the remark this answers is not in the document]
Comment (dukke) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: I wasn't saying to increase the padding in the right button graphic of the textfield. We can leave as it is..
We can adjust the padding on the right button instead so that the sum of the right button with the rightbutton graphic padding gets you the width you want.

[the remark this answers is not in the document]
Comment (dukke) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: Thanks for the suggestion of using `-fx-scale-x`
Though I think we could just use padding instead, makes it simpler. We can play with the padding of the right button and with the padding of the right button graphic.

[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: But how does the graphic actually behave for high-dpi displays?

[the remark this answers is not in the document]
Comment (dukke) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: I think we're probably not on the same page.
I probably didn't explain myself well.

On your previous comment you said you tried to change the padding on the right button graphic for the cross button. That was not what I suggested.
I'm saying leave the padding of the right button graphic for the cross button as it is but adjust the padding on the right button (not the right button graphic) if we want to make the right button width similar to the right button width of the password field. I think that's what you wanted and what we are now discussing here?

[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: I realised that and i committed a corrected version yesterday. See 756a7ed

[the remark this answers is not in the document]
Comment (dukke) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: Great! Thanks and thanks for your patience in discussing these issues.

Just one tiny small last suggestion before I approve and merge. Could we have it like this, instead of calling `textField.prefWidth(-1)` since we already know the right button height:

`layoutChiildren` method:
```
final double rightButtonHeight = h + textField.snappedTopInset() + textField.snappedBottomInset();
final double rightButtonWidth = rightButton.prefWidth(rightButtonHeight);
```

[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: I don't know what you mean. We never call `textField.prefWidth(-1)`

[the remark this answers is not in the document]
Comment (Exopandora) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: I already changed that in 1cd80a4

[the remark this answers is not in the document]
Comment (dukke) [reply FXSkins/src/main/java/impl/com/pixelduke/control/skin/TextFieldWithButtonSkin.java]: Yeah sorry, I meant `rightButton.prefWidth(-1);` 
( I meant to write `rightButton` but I wrote textField instead).

**Comment by dukke** [comment]
@Exopandora I've merged the PR.
Thanks a lot for your work! This is a fine addition to FXSkins! 👍👍

I'll try to integrate this ASAP into JMetro. I'll have to change JMetro to start using FXSkins and remove all skins present there.

Thanks again.

**Comment by dukke** [comment]
@Exopandora 
BTW as a token of recognition and appreciation for people who contribute to the project, like you, I'm going to start to prioritize issues, etc they post on the issue tracker.

**Comment by Exopandora** [comment]
Thank you very much!

**Comment by dukke** [comment]
Hi @Exopandora 

I just mentioned your PR in the latest JMetro announcement: https://pixelduke.com/2021/05/17/jmetro-version-11-6-15-released/

Thanks!

**Comment by Exopandora** [comment]
Thanks!
